cluster-autoscaler/cloudprovider/openshiftmachineapi/machineapi_controller_test.go

```diff
@@ -371,3 +394,533 @@ func TestMachinesInMachineSet(t *testing.T) {
+		},
+	}, {
+		description: "errors because maxSize is invalid",
+		annotations: map[string]string{
+			nodeGroupMinSizeAnnotationKey: "0",
+			nodeGroupMaxSizeAnnotationKey: "-1",
+		},
+	}, {
+		description: "errors because minSize > maxSize",
+		annotations: map[string]string{
+			nodeGroupMinSizeAnnotationKey: "1",
+			nodeGroupMaxSizeAnnotationKey: "0",
+		},
+	}, {
+		description: "errors because maxSize < minSize",
+		annotations: map[string]string{
+			nodeGroupMinSizeAnnotationKey: "1",
+			nodeGroupMaxSizeAnnotationKey: "0",
+		},
+	}, {
+		annotations: map[string]string{
+			nodeGroupMaxSizeAnnotationKey: "10",
+		},
+		count: 1,
+	}, {
+		annotations: map[string]string{
+			nodeGroupMinSizeAnnotationKey: "1",
+			nodeGroupMaxSizeAnnotationKey: "10",
+		},
+		count: 1,
+	}} {
+		machineSet := &v1beta1.MachineSet{
+			TypeMeta: v1.TypeMeta{
+				Kind: "MachineSet",
+			},
+			ObjectMeta: v1.ObjectMeta{
+				Name:        fmt.Sprintf("machineset-%d", i),
+				Namespace:   "test-namespace",
+				Annotations: tc.annotations,
+			},
+		}
+
+		controller, stop := mustCreateTestController(t, testControllerConfig{
+			machineObjects: []runtime.Object{
+				machineSet,
+			},
+		})
+		defer stop()
+
+		nodegroups, err := controller.nodeGroups()
+		if tc.count == 0 && err == nil {
+			t.Fatalf("expected an error")
+		}
+
+		if l := len(nodegroups); l != tc.count {
+			t.Errorf("expected %v, got %v", tc.count, l)
+		}
+	}
+}
+
+func TestControllerNodeGroupForNodeWithMissingNode(t *testing.T) {
+	machine := &v1beta1.Machine{
+		TypeMeta: v1.TypeMeta{
+			Kind: "Machine",
+		},
+		ObjectMeta: v1.ObjectMeta{
+			Name:      "machine",
+			Namespace: "test-namespace",
+			OwnerReferences: []v1.OwnerReference{{
+				Kind: "MachineSet",
+				UID:  "ec21c5fb-a3d5-a45f-887b-6b49aa8fc218",
+				Name: "testMachineSet",
+			}},
+		},
+	}
+
+	machineSet := &v1beta1.MachineSet{
+		TypeMeta: v1.TypeMeta{
+			Kind: "MachineSet",
+		},
+		ObjectMeta: v1.ObjectMeta{
+			Name:      "machineset",
+			Namespace: "test-namespace",
+			UID:       "ec21c5fb-a3d5-a45f-887b-6b49aa8fc218",
+		},
+	}
+
+	controller, stop := mustCreateTestController(t, testControllerConfig{
+		machineObjects: []runtime.Object{
+			machine,
+			machineSet,
+		},
+	})
+	defer stop()
+
+	ng, err := controller.nodeGroupForNode(&apiv1.Node{
+		TypeMeta: v1.TypeMeta{
+			Kind: "Node",
+		},
+		ObjectMeta: v1.ObjectMeta{
+			Name: "node",
+		},
+	})
+
+	if err != nil {
+		t.Fatalf("unexpected error: %v", err)
+	}
+
+	if ng != nil {
+		t.Fatalf("unexpected nodegroup: %v", ng)
+	}
+}
+
+func TestControllerNodeGroupForNodeWithMissingMachineOwner(t *testing.T) {
+	node := &apiv1.Node{
+		TypeMeta: v1.TypeMeta{
+			Kind: "Node",
+		},
+		ObjectMeta: v1.ObjectMeta{
+			Name: "node",
+			Annotations: map[string]string{
+				machineAnnotationKey: "test-namespace/machine",
+			},
+		},
+		Spec: apiv1.NodeSpec{
+			ProviderID: "provider-id",
+		},
+	}
+
+	machine := &v1beta1.Machine{
```

Comment:
I would rather move the machine definition under function with a set of parameters so there is only once place where the object is actually defined.

Reply:
I'm not sure whether this would buy us much as we have the following variations: 

```go
	testMachine := &v1beta1.Machine{
		ObjectMeta: v1.ObjectMeta{
			Name:      "test",
			Namespace: "test-namespace",
		},
	}
```
```go
	testMachine := &v1beta1.Machine{
		TypeMeta: v1.TypeMeta{
			Kind: "Machine",
		},
		ObjectMeta: v1.ObjectMeta{
			Name: "worker-us-east-2c-p4zwl",
		},
	}
```
```go
	machine := &v1beta1.Machine{
		TypeMeta: v1.TypeMeta{
			Kind: "Machine",
		},
		ObjectMeta: v1.ObjectMeta{
			Name:      "machine",
			Namespace: "test-namespace",
			OwnerReferences: []v1.OwnerReference{{
				Kind: "MachineSet",
				UID:  "ec21c5fb-a3d5-a45f-887b-6b49aa8fc218",
				Name: "testMachineSet",
			}},
		},
	}
```
```go
	machine := &v1beta1.Machine{
		TypeMeta: v1.TypeMeta{
			Kind: "Machine",
		},
		ObjectMeta: v1.ObjectMeta{
			Name:      "machine",
			Namespace: "test-namespace",
		},
	}
```
```go
	machine := &v1beta1.Machine{
		TypeMeta: v1.TypeMeta{
			Kind: "Machine",
		},
		ObjectMeta: v1.ObjectMeta{
			Name:      "machine",
			Namespace: "test-namespace",
			OwnerReferences: []v1.OwnerReference{{
				Name: machineSet.Name,
				Kind: machineSet.Kind,
				UID:  machineSet.UID,
			}},
		},
	}
```
If we have to pass parameters to the helper function we would probably have to pass structs to convey what goes where which leads us to_almost_ becomes the same thing. For sure, `TypeMeta` is common (and repeated).

Arguably, having everything local to the function under test means you don't have to jump around.